### PR TITLE
Staging archiver

### PR DESCRIPTION
**Pull request by eerkaijun**
@JoeHowarth @tongzhi1998 opening a draft PR so it's easier for me to leave some comments/questions. feel free to close this off if another branch is the canonical one

### CODE REVIEW

File: monad-archive/src/archive_reader.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        table: String,
+        region: Option<String>,
+        concurrency: usize,
+        metrics: &Metrics,
+    ) -> ArchiveReader {
+        let sdk_config = get_aws_config(region).await;
+        ArchiveReader {
+            s3: S3Archive::new(S3Bucket::new(bucket, &sdk_config, metrics.clone())),
+            dynamodb: DynamoDBArchive::new(table, &sdk_config, concurrency, metrics.clone()),
+        }
```
Comment (eerkaijun) [inline monad-archive/src/archive_reader.rs:35]: just to double confirm - will s3 and dynamodb always run in the same region?

File: monad-archive/src/archive_reader.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    dynamodb: DynamoDBArchive,
+}
+
+impl ArchiveReader {
+    pub async fn new(
+        bucket: String,
+        table: String,
+        region: Option<String>,
+        concurrency: usize,
+        metrics: &Metrics,
```
Comment (eerkaijun) [inline monad-archive/src/archive_reader.rs:29]: can this be made `Option<Metrics>` instead of `Metrics`?

File: monad-archive/src/dynamodb.rs [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+                })
+            })
+            .collect::<Vec<Option<TxIndexedData>>>();
+        Ok(output)
+    }
+
+    pub async fn get_txdata(&self, key: impl Into<String>) -> Result<Option<TxIndexedData>> {
+        self.batch_get_txdata(&[key.into()])
+            .await
+            .map(|mut v| v.remove(0))
```
Comment (eerkaijun) [inline monad-archive/src/dynamodb.rs]: `v.remove(0)` crashes the program if `key` (the tx hash) does not exist in archive, we should error handle here

File: monad-archive/src/dynamodb.rs [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+    ) -> Result<()>;
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Default, RlpEncodable, RlpDecodable)]
+pub struct TxIndexedData {
+    pub block_num: u64,
+    pub tx: TransactionSigned,
+    pub trace: Vec<u8>,
+    pub receipt: ReceiptWithBloom,
+}
```
Comment (eerkaijun) [inline monad-archive/src/dynamodb.rs]: for RPC we will need to know the `tx_index` of the transaction too (i.e. the position of the transaction in the block). is there a way we can store that too

File: monad-archive/src/archive_reader.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        self.s3.get_block_by_number(block_num).await
+    }
+
+    pub async fn get_block_receipts(&self, block_number: u64) -> Result<Vec<ReceiptWithBloom>> {
+        self.s3.get_block_receipts(block_number).await
+    }
+
+    pub async fn get_block_traces(&self, block_number: u64) -> Result<Vec<Vec<u8>>> {
+        self.s3.get_block_traces(block_number).await
+    }
```
Comment (eerkaijun) [inline monad-archive/src/archive_reader.rs:71]: what would be the best way to get block traces by hash? would you recommend first calling `get_block_by_hash` in archive then `get_block_traces`?

Comment (tongzhi1998) [reply monad-archive/src/archive_reader.rs:71]: Yes, I think that's the best way.

File: monad-archive/src/archive_reader.rs (hunk 1 above)
Comment (JoeHowarth) [reply monad-archive/src/archive_reader.rs:35]: yes

**Pull request closed without merging**